os/fs/driver/mtd/smart.c

```diff
@@ -5453,10 +5455,36 @@ static int smart_journal_recovery(FAR struct smart_struct_s *dev, journal_log_t
 
 	fvdbg("address : %u\n", address);
 	type = GET_JOURNAL_TYPE(log->status);
-	/* TODO Recovery Logic Will be added, after recovery we should update dev structure too(cf.freecount) */
 	switch (type) {
 	case SMART_JOURNAL_TYPE_COMMIT:
+		/* Journal is valid, so we will verify written data on data sector with crc of MTD header in journal */
+		memcpy(dev->rwbuffer, &log->mtd_header, mtd_size);
+		ret = MTD_READ(dev->mtd, psector * dev->sectorsize + mtd_size, dev->sectorsize - mtd_size, (FAR uint8_t *)&dev->rwbuffer[mtd_size]);
+		if (ret != dev->sectorsize - mtd_size) {
+			fdbg("Read from data sector error : %d\n", ret);
+			return -EIO;
+		}
 		
+		ret = smart_validate_crc(dev);
+		/* If CRC is not same, then release data sector and checkout journal */
+		if (ret != OK) {
+			fdbg("Invalid data, release dada sector & journal!\n");
+#if CONFIG_SMARTFS_ERASEDSTATE == 0xFF
+			status &= ~SMART_STATUS_RELEASED;
+#else
+			status |= SMART_STATUS_RELEASED;
+#endif
+			offset = psector * dev->sectorsize + offsetof(struct smart_sect_header_s, status);
+			ret = MTD_WRITE(dev->mtd, offset, 1, &status);
+			if (ret != 1) {
+				fdbg("Release Invalid data sector Error ret : %d target sector : %d\n", ret, psector);
```

Comment:
Considering line #5499, this debug message format should follow the same format.

Reply:
Yeah I considered like you but it is different type of situation.
Also working flow of smart_journal_checkout_process is not clear for now, I will clean it next PR

Reply:
I will update it in this PR.